design-proposals/vm-backend-storage.md

```diff
@@ -0,0 +1,73 @@
+# Overview
+Today, anything that happens on the host side of VMIs (i.e. virt-launcher) is transient.  
+Any change a VMI makes outside of its own file-system is lost after a shutdown.  
+Such changes include EFI firmware settings and information stored inside the TPM.  
+This design provides a way to persist small VM-specific files used by virt-launcher.
+
+## Motivation
+Though we now support exposing an emulated TPM device to guests (vTPM), using the device in any meaningful way would require persisting data across reboots.  
+As we do not currently have a way to do that, all data stored by the VMI into the vTPM is lost after a shutdown (a soft reboot keeps the same virt-launcher, avoiding losses).
+
+## Goals
+- Being able to persist modifications of various emulated devices across reboots for a given VM
+- Enable live migration of VMs that use backend storage
+- Avoid consuming large amounts of storage
+- Prevent virt-launcher for VM X from seeing data that belongs to VM Y
+
+## Non Goals
+- Persisting everything. Some information that would usually persist on a non-ephemeral host just don't matter to us.
+- Supporting user modifications of the backend storage. Users whishing to inject their own files should do so at their own risk.
+
+## Definition of Users
+This feature is intended for end-users who wish to preserve the state of some emulated device across reboots.
+
+## User Stories
+As a VM user, I want to be able to store the encryption key of my VMs partitions inside the vTPM (example: BitLocker)
+
+## Repos
+kubevirt/kubevirt
+
+# Design
+The storage used for this would be 1 PVC per VM, and would be created on first start of a VM that needs backend storage.  
+This PVC would be RWO, about 2 MiB in size (still TBD), and use [csi-hostpath](https://github.com/kubevirt/hostpath-provisioner).  
```

Comment:
This point is not clear to me. While we can copy most of the data to the new PVC on the migration, we simply would be limited by storage topology on restarts. I think we need to express these constrain or choose a different path.

Reply:
You lost me here... What is the concern about restarts?

Reply:
I think Alexander explained it well https://github.com/kubevirt/community/pull/170#discussion_r841923983 . Basically, RWO doesn't guarantee that you can attach the volume to another node. Therefore the volume will be available only on the particular node and the VMI will need to be scheduled there when you stop-start the VM.

Reply:
Right, but after that, Alice suggested creating new PVCs on migration, which I think is the way to go.
That way we never have to mount it in 2 places at once. I guess I did not make that obvious enough in the document, will review.

Reply:
I'm not sure I understand the question/concern... Once a VM is stopped, its VMI is gone, the PVC is not in use anymore and it can be used by the new VMI next start, even if it's just RWO, no?

Reply:
Certain RWO PVC once their are bound they cannot be attached to another node. This is depends on the storage provider you are using. Then, the VMs that uses that PVC can run only on the node where the PVC has been bound. However, this is a general issue for all the cases where the VM has a RWO PVC

Reply:
Exactly as Alice pointed out. I don't understand why we should take this limitation rather than require RWX. I would argue we can have simpler logic too if we require RWX.